File: tests/templates/WatchExtension/InterfaceController.fs
namespace monotouchtestWatchKitExtension

open System
open System.Collections.Generic
open System.Threading

open WatchKit
open Foundation

open MonoTouch.NUnit.UI

[<Register ("InterfaceController")>]
type InterfaceController (handle: IntPtr) =
    inherit WKInterfaceController (handle)

    let mutable runner = Unchecked.defaultof<WatchOSRunner>
    let mutable running = false

    [<Outlet ("lblStatus")>]
    member val lblStatus = Unchecked.defaultof<WKInterfaceLabel> with get, set

    [<Outlet ("lblSuccess")>]
    member val lblSuccess = Unchecked.defaultof<WKInterfaceLabel> with get, set

    [<Outlet ("lblFailed")>]
    member val lblFailed = Unchecked.defaultof<WKInterfaceLabel> with get, set

    [<Outlet ("lblIgnInc")>]
    member val lblIgnInc = Unchecked.defaultof<WKInterfaceLabel> with get, set

    [<Outlet ("cmdRun")>]
    member val cmdRun = Unchecked.defaultof<WKInterfaceButton> with get, set

    member this.LoadTests () =
        let excludeCategories =
            [|
                "MobileNotWorking"
                "NotOnMac"
                "NotWorking"
                "ValueAdd"
                "CAS"
                "InetAccess"
                "NotWorkingLinqInterpreter"
                "RequiresBSDSockets"
                "BitcodeNotSupported"
            |]
        runner <- new WatchOSRunner ()
        runner.ExcludedCategories <- new HashSet<string> (excludeCategories :> IEnumerable<string>)
        let tp = this.GetType ()
        runner.Add (tp.Assembly)
        ThreadPool.QueueUserWorkItem (fun v ->
            runner.LoadSync ()
            this.BeginInvokeOnMainThread (fun x ->
                this.lblStatus.SetText (String.Format ("{0} tests", runner.TestCount))
                this.RenderResults ()
                this.cmdRun.SetEnabled (true)
                this.cmdRun.SetHidden (false)

                runner.AutoRun ()
            )
        )
        |> ignore

    override this.Awake (context: NSObject) =
        base.Awake (context)
        this.BeginInvokeOnMainThread (fun x ->
            this.LoadTests ()
            ()
        )

    member this.RenderResults () =
        if runner.TestCount > 0 then
            this.lblSuccess.SetText (String.Format ("P: {0}/{1} {2}%", runner.PassedCount, runner.TestCount, 100 * runner.PassedCount / runner.TestCount))
            this.lblFailed.SetText (String.Format ("F: {0}/{1} {2}%", runner.FailedCount, runner.TestCount, 100 * runner.FailedCount / runner.TestCount))
            this.lblIgnInc.SetText (String.Format ("I: {0}/{1} {2}%", (runner.IgnoredCount + runner.InconclusiveCount), runner.TestCount, 100 * (runner.IgnoredCount + runner.InconclusiveCount) / runner.TestCount))

            if running && runner.PassedCount > 0 then
                if runner.FailedCount = 0 then
                    this.lblSuccess.SetTextColor (UIKit.UIColor.Green);
                    this.lblStatus.SetTextColor (UIKit.UIColor.Green);
                    this.lblStatus.SetText ("Success");
                
                if runner.FailedCount > 0 then
                    this.lblFailed.SetTextColor (UIKit.UIColor.Red);
                    this.lblStatus.SetTextColor (UIKit.UIColor.Red);
                    this.lblStatus.SetText ("Failed");

    member this.RunTests () = 
        if running then
            printf "Already running"
        else
            let running = true
            this.cmdRun.SetEnabled (false)
            this.lblStatus.SetText ("Running")
            this.BeginInvokeOnMainThread (fun v ->
                runner.Run ()
                this.lblStatus.SetText ("Done")
                let running = false
                this.RenderResults ()
            )

    [<Action ("runTests:")>]
    member this.RunTests (obj: NSObject) =
        this.RunTests ()

File: tests/templates/WatchExtension/InterfaceController.cs
using System;
using System.Collections.Generic;
using System.Threading;

using WatchKit;
using Foundation;

using MonoTouch.NUnit.UI;

public static partial class TestLoader {
	static partial void AddTestAssembliesImpl (BaseTouchRunner runner);

	public static void AddTestAssemblies (BaseTouchRunner runner)
	{
		AddTestAssembliesImpl (runner);
	}
}

namespace monotouchtestWatchKitExtension {
	[Register ("InterfaceController")]
	public partial class InterfaceController : WKInterfaceController {
		WatchOSRunner runner;
		bool running;

		[Action ("runTests:")]
		partial void RunTests (NSObject obj);

		[Outlet ("lblStatus")]
		WatchKit.WKInterfaceLabel lblStatus { get; set; }

		[Outlet ("lblSuccess")]
		WatchKit.WKInterfaceLabel lblSuccess { get; set; }

		[Outlet ("lblFailed")]
		WatchKit.WKInterfaceLabel lblFailed { get; set; }

		[Outlet ("lblIgnInc")]
		WatchKit.WKInterfaceLabel lblIgnInc { get; set; }

		[Outlet ("cmdRun")]
		WatchKit.WKInterfaceButton cmdRun { get; set; }

		static InterfaceController ()
		{
			ObjCRuntime.Runtime.MarshalManagedException += (object sender, ObjCRuntime.MarshalManagedExceptionEventArgs args) => {
				Console.WriteLine ("Managed exception: {0}", args.Exception);
			};
			ObjCRuntime.Runtime.MarshalObjectiveCException += (object sender, ObjCRuntime.MarshalObjectiveCExceptionEventArgs args) => {
				Console.WriteLine ("Objective-C exception: {0}", args.Exception);
			};
		}

		public InterfaceController (IntPtr handle) : base (handle)
		{
		}

		public override void Awake (NSObject context)
		{
			base.Awake (context);

			BeginInvokeOnMainThread (LoadTests);
		}

		void LoadTests ()
		{
			var excludeCategories = new [] {
				"MobileNotWorking",
				"NotOnMac",
				"NotWorking",
				"ValueAdd",
				"CAS",
				"InetAccess",
				"NotWorkingLinqInterpreter",
				"RequiresBSDSockets",
				"BitcodeNotSupported",
			};
			runner = new WatchOSRunner ();
			runner.ExcludedCategories = new HashSet<string> (excludeCategories);
			runner.Add (GetType ().Assembly);
			TestLoader.AddTestAssemblies (runner);
			ThreadPool.QueueUserWorkItem ((v) => {
				runner.LoadSync ();
				BeginInvokeOnMainThread (() => {
					lblStatus.SetText (string.Format ("{0} tests", runner.TestCount));
					RenderResults ();
					cmdRun.SetEnabled (true);
					cmdRun.SetHidden (false);

					runner.AutoRun ();
				});
			});
		}

		void RunTests ()
		{
			if (running) {
				Console.WriteLine ("Already running");
				return;
			}
			running = true;
			cmdRun.SetEnabled (false);
			lblStatus.SetText ("Running in background");

			var timer = NSTimer.CreateRepeatingScheduledTimer (TimeSpan.FromSeconds (1), (v) => RenderResults ());
			var runnerThread = new Thread (() => {
				runner.Run ();

				InvokeOnMainThread (() => {
					cmdRun.SetEnabled (true);
					lblStatus.SetText ("Done");
					running = false;
					timer.Dispose ();
					RenderResults ();
				});
			}) {
				IsBackground = true,
			};
			runnerThread.Start ();
		}

		void RenderResults ()
		{
			if (runner.TestCount == 0)
				return;

			lblSuccess.SetText (string.Format ("P: {0}/{1} {2}%", runner.PassedCount, runner.TestCount, 100 * runner.PassedCount / runner.TestCount));
			lblFailed.SetText (string.Format ("F: {0}/{1} {2}%", runner.FailedCount, runner.TestCount, 100 * runner.FailedCount / runner.TestCount));
			lblIgnInc.SetText (string.Format ("I: {0}/{1} {2}%", (runner.IgnoredCount + runner.InconclusiveCount), runner.TestCount, 100 * (runner.IgnoredCount + runner.InconclusiveCount) / runner.TestCount));

			if (running == false && runner.PassedCount > 0) {
				if (runner.FailedCount == 0) {
					lblSuccess.SetTextColor (UIKit.UIColor.Green);
					lblStatus.SetTextColor (UIKit.UIColor.Green);
					lblStatus.SetText ("Success");
				}
				if (runner.FailedCount > 0) {
					lblFailed.SetTextColor (UIKit.UIColor.Red);
					lblStatus.SetTextColor (UIKit.UIColor.Red);
					lblStatus.SetText ("Failed");
				}
			}
		}

		partial void RunTests (NSObject obj)
		{
			RunTests ();
		}
	}
}
